### PR TITLE
Fix building of discriminated union where field_validator wraps a Literal

### PR DESCRIPTION
**Pull request by adriangb**
skip change file check

Closes https://github.com/pydantic/pydantic/issues/6525

Selected Reviewer: @samuelcolvin

### CODE REVIEW

File: tests/test_discriminated_union.py [hunk 1]
```diff
@@ -1232,3 +1232,72 @@ class TestModel(BaseModel):
...
+            return 'sparrow'
+
+    Animal = Annotated[
+        CatModel | SparrowModel,
+        Field(discriminator='name'),
+    ]
+
+    ta = TypeAdapter(Animal)
+
+    assert ta.validate_python({'name': 'kitty'}) == CatModel(name='cat')
```
Comment (ybressler) [inline tests/test_discriminated_union.py:1261]: What if both models have a common value – I would expect the firstmost defined object to return. Worth testing?

File: tests/test_discriminated_union.py [hunk 2]
```diff
@@ -1232,3 +1232,72 @@ class TestModel(BaseModel):
...
+            'mapping': {
+                'birdie': '#/$defs/SparrowModel',
+                'cat': '#/$defs/CatModel',
+                'kitty': '#/$defs/CatModel',
+                'sparrow': '#/$defs/SparrowModel',
+            },
+            'propertyName': 'name',
+        },
+        'oneOf': [{'$ref': '#/$defs/CatModel'}, {'$ref': '#/$defs/SparrowModel'}],
+    }
```
Comment (ybressler) [inline tests/test_discriminated_union.py:1303]: Thorough! But a bit tedious to write + replicate. Any way to simplify this part of the code?

If what we care about are the `$defs` then maybe:
```
    assert ta.json_schema() == {
        '$defs': {
            CatModel.json_schema(),
            SparrowModel.json_schema(),
        },
        'discriminator': {
            'mapping': {
                'birdie': '#/$defs/SparrowModel',
                'cat': '#/$defs/CatModel',
                'kitty': '#/$defs/CatModel',
                'sparrow': '#/$defs/SparrowModel',
            },
            'propertyName': 'name',
        },
        'oneOf': [{'$ref': '#/$defs/CatModel'}, {'$ref': '#/$defs/SparrowModel'}],
    }
```

**Review by ybressler — COMMENTED** [review]
Code looks good! Left a few inquisitive Q's :)

File: tests/test_discriminated_union.py (hunk 1 above)
Comment (adriangb) [reply tests/test_discriminated_union.py:1261]: Yes indeed! But that's already tested elsewhere (in this file, I can look for the right line tomorrow; if it's not us should be!)

Comment (adriangb) [reply tests/test_discriminated_union.py:1261]: (To be clear, I believe the actual behavior is to error during schema building)

File: tests/test_discriminated_union.py (hunk 2 above)
Comment (adriangb) [reply tests/test_discriminated_union.py:1303]: Indeed that'd be less verbose, but the `insert_assert` utility does snapshot testing so this is not hand written and doesn't need to be updated by hand 😀

**Comment by adriangb** [comment]
please review cc @dmontagu

**Review by ybressler — APPROVED** [review]
Great work.

**Comment by adriangb** [comment]
Closing in favor of #6570

**Pull request closed without merging**